src/shogun/optimization/FirstOrderMinimizer.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2015 Wu Lin
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+#ifndef FIRSTORDERMINIMIZER_H
+#define FIRSTORDERMINIMIZER_H
+#include <shogun/lib/config.h>
+#include <shogun/optimization/FirstOrderCostFunction.h>
+#include <shogun/optimization/MinimizerContext.h>
+#include <shogun/optimization/Penalty.h>
+namespace shogun
+{
+
+/** @brief The first order minimizer base class.
+ *
+ * This class gives the interface of a first-order gradient-based unconstrained minimizer
+ *
+ * This kind of minimizers will find optimal target variables based on gradient information wrt target variables.
+ * For example, the gradient descend method is a minimizer.
+ *
+ * A minimizer need the following things as input:
+ * a supported cost function object 
+ * a penalty object if regularization is enabled
+ * a context object to restore mutable variables if deserialization is actived
+ *
+ */
+class FirstOrderMinimizer
+{
+public: 
+	/** Default constructor */
+	FirstOrderMinimizer()
+	{
+		init();
+	}
+
+	/** Constructor
+	 * @param fun cost function (user have to manully delete the pointer)
```

Comment:
What do you think about having a Shogun base class without serialisation and all the stuff. ONLY reference counting.
Then all your classes could be subclasses of that one and users would not have to do things.

See #2880